File: golang/binary_search_tree/min_absolute_diff_bst.go
package main

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func getMinimumDifference(root *TreeNode) int {
	minDiff := math.MaxInt32

	var prev *int

	var inorder func(node *TreeNode)
	inorder = func(node *TreeNode) {
		if node == nil {
			return
		}

		// Traversal ke kiri
		inorder(node.Left)

		// Proses node saat ini
		if prev != nil {
			minDiff = min(minDiff, node.Val-*prev)
		}
		prev = &node.Val

		// Traversal ke kanan
		inorder(node.Right)
	}

	inorder(root)
	return minDiff
}

File: python/binary_search_tree/min_absolute_diff_bst.py
import math


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def getMinimumDifference(self,root: TreeNode) -> int:
        min_diff = math.inf
        prev = None

        def inorder(node: TreeNode):
            nonlocal min_diff, prev

            if node is None:
                return

            inorder(node.left)

            if prev is not None:
                min_diff = min(min_diff, node.val - prev)
            prev = node.val

                    # Traversal ke kanan
            inorder(node.right)


        inorder(root)

        return min_diff
